mp/src/game/shared/momentum/mom_gamemovement.cpp

```diff
@@ -1283,31 +1283,77 @@ void CMomentumGameMovement::CategorizePosition()
-                                 1.0 + sv_bounce.GetFloat() * (1 - player->m_surfaceFriction));
+                    trace_t pmFall;
+
+                    // Don't simulate the fall to the ground if the player is already directly on it
+                    if (pm.fraction != 0.0f)
+                    {
+                        // Simulate the fall next tick
+                        Vector endFall;
+                        VectorMA(mv->GetAbsOrigin(), gpGlobals->frametime, vecNextVelocity, endFall);
+
+                        TracePlayerBBox(mv->GetAbsOrigin(), endFall, PlayerSolidMask(), COLLISION_GROUP_PLAYER_MOVEMENT, pmFall);
+                    }
+                    else if (m_pPlayer->GetLastCollisionTick() == gpGlobals->tickcount)
+                    {
+                        // Pretend we are starting the edgebug from where we hit the surface and then simulate a max
+                        // distance edgebug to ensure consistency (see note on the sliding simulation)
+                        pmFall = m_pPlayer->GetLastCollisionTrace();
+                    }
+                    else
+                    {
+                        // Fallback if the fraction is zero, but the player hasn't collided this tick for some reason
+                        pmFall = pm;
+                    }
+
+                    if (!pmFall.DidHit())
+                    {
+                        bGrounded = false; // Would miss the ground the next tick
+                    }
+                    else
+                    {
+                        ClipVelocity(vecNextVelocity, pmFall.plane.normal, vecNextVelocity, 1.0f);
+
+                        // Simulate the sliding on the ground (notice that we are using the full frametime to allow
+                        // consistent max distance edgebugs)
+                        Vector endSlide;
+                        VectorMA(pmFall.endpos, gpGlobals->frametime, vecNextVelocity, endSlide);
+
+                        trace_t pmSlide;
+                        TracePlayerBBox(mv->GetAbsOrigin(), endSlide, PlayerSolidMask(), COLLISION_GROUP_PLAYER_MOVEMENT, pmSlide);
+
+                        // Check if there is ground under the player (otherwise it would be an edgebug)
+                        trace_t pmGround;
+                        TryTouchGround(pmSlide.endpos, pmSlide.endpos - Vector(0.0f, 0.0f, flOffset), GetPlayerMins(),
+                                       GetPlayerMaxs(), MASK_PLAYERSOLID, COLLISION_GROUP_PLAYER_MOVEMENT, pmGround);
+
+                        if (!pmGround.DidHit())
+                            bGrounded = false; // Would edgebug next tick
+                    }
                 }
 
-                // Set ground entity if the player is not going to slide on the ramp next tick
-                if (rampVelocity[2] <= NON_JUMP_VELOCITY)
+                ClipVelocity(vecNextVelocity, pm.plane.normal, vecNextVelocity, 1.0f);
```

Comment:
What happened to checking the surface normal Z component? This seems to remove the original functionality of checking if the surface is a jumpable slope or a ramp, this could negatively affect surf or bhop

Reply:
We don't need to check it since this code wouldn't run if it wasn't something the player could stand on.

Reply:
I gotcha, just read the initial if/else and saw it catches it when the normal is `< 0.7` anyways.